Improve location permission handling

allroundexperts opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
This PR improves the handling of location permissions.

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/37736
PROPOSAL: N/A


### Tests

**1) Prompted for permission, prompt rejected**
- _Setup: Location permission DISABLED / Location feature OFF_
- Create a money request from another user via the scan receipt flow
- Submit it
- In-app 'Allow location access' Modal should appear
- Tap 'Not now'
- The expense should submit

**2) Prompted for permission, prompt accepted, OS prompt rejected**
- _[iOS/Android/Web ONLY]_
- _Setup: Location permission DISABLED / Location feature OFF_
- Create a money request from another user via the scan receipt flow and submit it
- In-app 'Allow location access' Modal should appear
- Tap 'Continue'
- The OS permission prompt should appear
- _[iOS/Android ONLY from here on]_
- Reject it
- Keep repeating the above steps and rejecting the OS permission prompt
- (1-3, THE NUMBER OF TIMES YOU NEED TO DECLINE DEPENDS ON OS)
- Eventually the OS permission prompt will stop showing, instead our 'Allow camera access' modal should be displayed

**3) Prompted for permission, prompt rejected**
- _Setup: Location permission ENABLED / Location feature ON_
- Create a money request from another user via the scan receipt flow
- Submit it
- No modals should appear

**4) Prompted for permission, prompt rejected**
- _[iOS/Android ONLY]_
- _Setup: Location permission ENABLED / Location feature OFF_
- Create a money request from another user via the scan receipt flow
- Submit it
- You should be prompted by the OS to enable location/GPS
- Select any option
- The expense should succeed


### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
N/A

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
8. Upload an image via copy paste
9. Verify a modal appears displaying a preview of that image
--->

Same as above

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/30054992/5ea51c5e-ae4d-4799-8eb1-e4c505d170f5

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/30054992/c1ddefa7-0f2b-4f59-809c-1aca726577bf

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/30054992/f78f5c2b-cb3b-4fdf-b3d8-e84c4fb3fae4

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/30054992/ac85ea61-d128-4065-bcdd-0007119662cf

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/30054992/fa57bbbb-4dc1-4330-8b25-c65f25645cd2

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/30054992/accd6540-87e1-46dc-998f-47087dd6aa20

</details>

allroundexperts committed:
temp feature

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts committed:
design ready

allroundexperts committed:
add logic for getting permission status

Julesssss commented on src/components/ConfirmContent.tsx:
```diff
@@ -51,15 +54,33 @@ type ConfirmContentProps = {
+    /** Icon height */
+    iconHeight?: number;
+
+    /** Should the icon be centered? */
+    shouldCenterIcon?: boolean;
```
Is there a reason we wouldn't want to centre it?

Julesssss commented on src/components/ConfirmContent.tsx:
```diff
@@ -139,17 +182,25 @@ function ConfirmContent({
 
                 {shouldStackButtons ? (
                     <>
+                        {shouldShowCancelButton && shouldReverseStackedButtons && (
+                            <Button
+                                style={[styles.mt4, styles.noSelect]}
+                                onPress={onCancel}
+                                large
+                                text={cancelText || translate('common.no')}
+                            />
+                        )}
```
The `shouldReverseStackedButtons` logic seems a bit complex. We could check with design that the button have to be ordered that way

allroundexperts replied:
That'd be good. @Expensify/design Can you please weigh in on if the buttons need to be ordered in the way they are currently?

<img width="357" alt="Screenshot 2024-04-09 at 2 11 42 PM" src="https://github.com/Expensify/App/assets/30054992/fd02057a-99a6-4782-b719-5fefc141560f">

shawnborton replied:
I think we want the green/success CTA at the bottom, based on @dannymcclain's mockups. From a code perspective I can't offer much - what exactly is the issue there?

Julesssss replied:
It's not a huge problem, but the code is complicated by three separate rendering conditions and it's more likely that future modifications to the code will lead to regressions. 

Though it's not a huge deal, so let's stick with the original design.

The better solution would be to refactor the buttons to a subcomponent, but there's no need to do that as part of this PR 👍

dannymcclain replied:
Gotcha. Yeah we ideally want the green `Continue` button to be the easiest button to reach/click, which is why it's on the bottom.

allroundexperts committed:
correct logic for web

allroundexperts commented:
This is working pretty good on all the platforms. I think I'll make this ready for review today.

Julesssss commented:
Nice, that would be great. Getting this in review today might let us get it merged before I go OOO on Friday 🤞

allroundexperts committed:
merge with main

allroundexperts committed:
lint fix

allroundexperts committed:
lint fix

allroundexperts committed:
fix: ios permission issue

allroundexperts committed:
merge with main

allroundexperts commented:
@Julesssss I was not able to create a PR yesterday because of a weird bug on iOS for `react-native-permissions` library (https://github.com/zoontek/react-native-permissions/issues/813). I've updated the library to a newer version and it fixed the issue. Just FYI.

allroundexperts committed:
Show error only when location is blocked

allroundexperts committed:
icon not showing for location

Julesssss commented:
@allroundexperts no worries. if it's alright with you I'd prefer to hold this for a week until I'm back. There are quite a few things to test here so I don't want to hand it over to another engineer. 

Lets open it up for review when I'm back and we'll find a C+ member who is around to get it reviewed quickly 👍

allroundexperts commented:
All good!

abdulrahuman5196 commented:
> @abdulrahuman5196 Is having limited availability for the next couple of weeks. @Julesssss Let's re-assign to someone who can review quicker.

I would be available to review when this PR comes back for review.

Julesssss commented:
Hey @allroundexperts I'm back. Can you please fix the conflicts and we can start to test this PR?

allroundexperts commented:
Cool cool. I'll resolve the conflicts today!

Julesssss commented:
> Cool cool. I'll resolve the conflicts today!

Thanks. Once done I'll create automated builds.

allroundexperts committed:
merge with main

allroundexperts committed:
remove play sound

allroundexperts commented:
@Julesssss I see that we added location permissions to the distance requests as well (It asks for permission as soon as we open the distance request tab). Do you know how we'd handle that case? Or should we just worry about it in a follow up issue?

allroundexperts commented:
I have resolved conflicts and this should be ready for testing.

allroundexperts committed:
lint fix

allroundexperts commented:
@shubham1206agra Can you post a video of test case failing?

allroundexperts committed:
merge with main

shawnborton commented:
In the screenshots above, how did you get the camera in this state where the capture button isn't centered?
![CleanShot 2024-04-30 at 09 02 02@2x](https://github.com/Expensify/App/assets/2319350/7165c147-a4ae-4ef4-9f8d-2e8d18b2b3d9)

shawnborton commented:
Actually that makes me wonder - are those screenshots all updated?

allroundexperts commented:
@shawnborton The screen recordings were taken about 3 weeks ago. I'll update them again.

dubielzyk-expensify commented:
I just tested the adhoc build and it completely crashed when I went through these steps:
1. Tapped (+) and Submit expense
2. Scanned receipt
3. Added @dannymcclain as the contact
4. Modal popped up successfully (👍 ) and I tapped Not now.
5. Deleted the expense
6. Ran through steps 1-4 again, but this time tapping allow, then "Allow once" (not the other option which says "Allow while using app".
7. App crashed and now won't reopen at all even after killing the app

dubielzyk-expensify commented:
The modal itself looks mostly great, but 2 things:
1. The illustration renders weirdly and looks broken:
2. And the X isn't tucked up in the top right corners as per the mocks (which is 20px from top right)

Screenshot vs mock:
![image](https://github.com/Expensify/App/assets/145416827/bba2b1aa-6f86-41c1-a074-efed3a90be92)


As for the illustration, I wonder if we need to reexport the asset. You run into this before @shawnborton or @dannymcclain ?

Julesssss reviewed (dismissed):
The Android flow looks good, and I can confirm the GPS coordinates were submitted correctly once I accepted the permission AND enabled fine location (GPS) on my device:

However, I also had the crash when submitting a regular, non camera scan expense

_(GH won't let me drop the screenshot, so see below in the issue)_

Julesssss commented:
![Screenshot 2024-05-01 at 10 33 12](https://github.com/Expensify/App/assets/10736861/3851493f-0d5b-4d49-9e53-30f7f91dcc76)

shawnborton commented:
> As for the illustration, I wonder if we need to reexport the asset. 

Hmm, can't hurt to run it through illustrator. I usually end up re-exporting via illustrator and in doing so, I usually try to get rid of all of the clipping masks and try to simplify the shapes quite a bit. Figma's SVG export adds a lot of shape complexity from my experience.

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts commented:
@shubham1206agra Web seems to work fine for me.

https://github.com/Expensify/App/assets/30054992/cb1c8efb-343c-49de-9d76-988907e5d776

Can you search your console for `[IOURequestStepConfirmation] getCurrentPosition failed` and see what it says?

allroundexperts committed:
fix: close icon margin top

allroundexperts commented:
> The modal itself looks mostly great, but 2 things:
> 
> 1. The illustration renders weirdly and looks broken:
> 2. And the X isn't tucked up in the top right corners as per the mocks (which is 20px from top right)
> 
> Screenshot vs mock: ![image](https://private-user-images.githubusercontent.com/145416827/327062278-bba2b1aa-6f86-41c1-a074-efed3a90be92.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ2MDYzOTgsIm5iZiI6MTcxNDYwNjA5OCwicGF0aCI6Ii8xNDU0MTY4MjcvMzI3MDYyMjc4LWJiYTJiMWFhLTZmODYtNDFjMS1hMDc0LWVmZWQzYTkwYmU5Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwMVQyMzI4MThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NTk4NzE1ODkxMjIwNzkyYmYxMmJhOTAyZTkzZGQ0Mjc3NzM2OTA5NzZlMjg2ZjQ3NjQ5OTFiZWM0ZDY5NmVmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.hmatkyB5fLsEAx844Ss5AeCiSLm7A0jfiL6942J3T0M)
> 
> As for the illustration, I wonder if we need to reexport the asset. You run into this before @shawnborton or @dannymcclain ?

Fixed the close icon issue.

<img width="408" alt="Screenshot 2024-05-02 at 4 34 18 AM" src="https://github.com/Expensify/App/assets/30054992/6888dcdf-996e-4857-a34b-d2cc36c0fb3d">

Can you please provide me the correct illustration and I'll add it.

allroundexperts commented:
> I just tested the adhoc build and it completely crashed when I went through these steps:
> 
> 1. Tapped (+) and Submit expense
> 2. Scanned receipt
> 3. Added @dannymcclain as the contact
> 4. Modal popped up successfully (👍 ) and I tapped Not now.
> 5. Deleted the expense
> 6. Ran through steps 1-4 again, but this time tapping allow, then "Allow once" (not the other option which says "Allow while using app".
> 7. App crashed and now won't reopen at all even after killing the app

@dubielzyk-expensify I am not able to reproduce this. Can you maybe post a video?

dubielzyk-expensify commented:
Might be something weird with the screenshot, but the X still looks a tiny bit off vertically:
![CleanShot 2024-05-02 at 13 33 55@2x](https://github.com/Expensify/App/assets/145416827/190d3ad7-1cdd-4a04-b4e7-8858c9d7edd8)

dubielzyk-expensify commented:
Here ya go. Now it just crashes altogether:

https://github.com/Expensify/App/assets/145416827/a04a7ee7-587d-47d7-9128-f0879f61bf9b

dannymcclain commented:
Coming from [this thread](https://github.com/Expensify/App/issues/41252#issuecomment-2090346813), I'm actually not sure this should even have a close button. Looks like we decided over there that click outside/escape is enough, and we don't need to included a dedicated close (x) button.

As for the illustration—no idea what's going on there. I followed the usual procedure when I initially exported it (Figma > Illustrator), but here's a fresh export. I tried to clean the shapes/paths up a little bit in Illustrator as Shawn suggested before saving this one out. Let's see if it makes a difference! 

[Globe with Receipt and Location Marker-2.svg.zip](https://github.com/Expensify/App/files/15189281/Globe.with.Receipt.and.Location.Marker-2.svg.zip)

shawnborton commented:
Yeah great shout on the thread about the close "x" - seems like we can kill that.

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts commented:
@shawnborton @dubielzyk-expensify `x` removed and added the new illustration.

<img width="420" alt="Screenshot 2024-05-02 at 8 21 46 PM" src="https://github.com/Expensify/App/assets/30054992/88abf4e8-bbb8-4382-92e0-8b964f5bd8dd">

allroundexperts committed:
update illustration and remove x

allroundexperts commented:
@Julesssss I am unable to reproduce the crash that @dubielzyk-expensify mentioned. Can you maybe re-trigger the build so that we can test again with the latest code?

Julesssss commented:
@allroundexperts I just ran a new build for us to test. In the meantime I can reliably reproduce a different crash with these steps on multiple accounts:
- Request money
- Choose a contact
- Choose Request money instead of scan receipt
- type any amount
- Select any person (I haven't tried workspace)
- Submit
- App crashes

allroundexperts commented:
> @allroundexperts I just ran a new build for us to test. In the meantime I can reliably reproduce a different crash with these steps on multiple accounts:
> 
> * Request money
> * Choose a contact
> * Choose Request money instead of scan receipt
> * type any amount
> * Select any person (I haven't tried workspace)
> * Submit
> * App crashes

On Android, @Julesssss?

dubielzyk-expensify commented:
I can't seem to reproduce the crash with the newest adhoc build. Everything else looks good to me now 👍

Julesssss commented:
> On Android, @Julesssss?

Hey Sibtain. I get the same crash on both Android and iOS (videos below). I managed to grab some logs today: 

![Screenshot 2024-05-03 at 09 49 06](https://github.com/Expensify/App/assets/10736861/c42cc11c-8c81-4a7e-b1a8-28a2818e0208)

**Android**

https://github.com/Expensify/App/assets/10736861/251810aa-7fb1-4e83-ab2a-9cff2f023043



**iOS**


https://github.com/Expensify/App/assets/10736861/321470f7-a4f7-477c-8131-0250be14912a

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts commented:
> > On Android, @Julesssss?
> 
> Hey Sibtain. I get the same [crash](https://github.com/Expensify/App/pull/39875#issuecomment-2090807591) on both Android and iOS (videos below). I also tested another AdHoc build and didn't get the crash. I managed to grab some logs today:
> 
> ![Screenshot 2024-05-03 at 09 49 06](https://private-user-images.githubusercontent.com/10736861/327701819-c42cc11c-8c81-4a7e-b1a8-28a2818e0208.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUwMTE1ODIsIm5iZiI6MTcxNTAxMTI4MiwicGF0aCI6Ii8xMDczNjg2MS8zMjc3MDE4MTktYzQyY2MxMWMtOGM4MS00YTdlLWIxYTgtMjhhMjgxOGUwMjA4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA2VDE2MDEyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiYWY0ZGJhMjg2MmVhMjIzN2JiYjFjMjA1ZGNkMTg4YzM0M2E1NjU1MzBiNzg0OWRkNjBjYmFlZjFmNzI4NzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MrJVWRN3dW2_fw0_NY5k_wyaRlDBuKBGmBPvmTHlz4g)
> 
> **Android**
> 
>  Android-Crash.mp4 
> **iOS**
> 
>  iOS-Crash.mp4

@Julesssss This doesn't seem to be related to this PR. I just merged main again and pushed. Can you trigger a new build and test again?

Julesssss commented:
@allroundexperts yep, trying again as the Android build failed.

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts committed:
fix: crash when manually requesting money

allroundexperts commented:
@Julesssss @shubham1206agra Crash is fixed.

allroundexperts committed:
fix: ts errors

allroundexperts commented:
Fixed @shubham1206agra

Julesssss reviewed (commented):
Thanks for fixing, new builds are in progress

allroundexperts commented:
@shubham1206agra Can you check locally?

Julesssss commented:
I can still reproduce on Android and I also see it on that test web [build](https://39875.pr-testing.expensify.com/). That should be easier to test and fix 👍 

![Screenshot 2024-05-09 at 14 17 24](https://github.com/Expensify/App/assets/10736861/2c9922a9-0d07-4d03-bd24-30c8aa9b0b59)

Julesssss commented:
It is odd that this isn't reproducible locally though 😖 

I'm going to try submitting a PR to resolve the crash, then you can try merging main again.

Julesssss reviewed (dismissed):
@allroundexperts after trying to resolve this locally I think this has been introduced by the `selectedParticipantsRef` [change](https://github.com/Expensify/App/pull/39875/files#diff-5e5fc302dd5db871d26e30f810bc0606beccb7f50ae6ed6535cd93aaee98c280R87). The PR I was working on won't help because we rely on many other properties of the null participant object.

I haven't located the full reproduction steps, but can you take another look please? The loc seems suspicious because the crash is due to the selectedParticipant being null.

Julesssss commented on src/pages/iou/request/step/IOURequestStepConfirmation.tsx:
```diff
@@ -535,6 +545,21 @@ function IOURequestStepConfirmation({
+    // To prevent the component from rendering with the wrong currency, we show a loading indicator until the correct currency is set.
+    const isLoading = !!transaction?.originalCurrency;
+
+    const onConfirm = (listOfParticipants: Participant[]) => {
+        selectedParticipantsRef.current = listOfParticipants;
```
Is `listOfParticipants` getting set correctly here?

Julesssss commented on src/pages/iou/request/step/IOURequestStepConfirmation.tsx:
```diff
@@ -80,6 +83,9 @@ function IOURequestStepConfirmation({
     const {windowWidth} = useWindowDimensions();
     const {isOffline} = useNetwork();
+    const [startLocationPermissionFlow, setStartLocationPermissionFlow] = useState(false);
+    const selectedParticipantsRef = useRef<Participant[]>([]);
```
Is this correct? `useRef<Participant[]>([]);`

Julesssss commented on src/pages/iou/request/step/IOURequestStepConfirmation.tsx:
```diff
@@ -315,15 +323,15 @@ function IOURequestStepConfirmation({
                     .filter((accountID: string): boolean => (transaction?.splitShares?.[Number(accountID)]?.amount ?? 0) > 0)
                     .map((accountID) => Number(accountID));
-                splitParticipants = selectedParticipants.filter((participant) => participantsWithAmount.includes(participant.accountID ?? -1));
+                splitParticipants = selectedParticipantsRef.current.filter((participant) => participantsWithAmount.includes(participant.accountID ?? -1));
```
Are we sure this filter is working correctly?

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts commented:
@Julesssss I still can't reproduce the crash. Can you by any chance, run the latest code locally and send a screen recording?

Julesssss commented:
I can't reproduce locally because I don't know the full steps, but this occurs on builds pointing to prod and we can't merge this PR as is. I'm afraid the best I can do is point to the code which is the most likely cause and share the [stack trace.](https://github.com/Expensify/App/pull/39875#issuecomment-2102646525)

Is there an alternative solution that doesn't depend on the [new](https://github.com/Expensify/App/pull/39875/files/da3dc4678d64ab02065ce12018cc7add391d6252#r1599619451) `useRef<Participant>`?

Julesssss commented:
The stack trace is pretty clear in that we are not correctly passing a participant into the component [here](https://github.com/Expensify/App/pull/39875/files#r1599620046) in some cases.

allroundexperts commented:
> Is there an alternative solution that doesn't depend on the [new](https://github.com/Expensify/App/pull/39875/files/da3dc4678d64ab02065ce12018cc7add391d6252#r1599619451) `useRef<Participant>`

The alternative was to store the participants in the state, which actually causes crash every time because the state is not updated when the confirm handler is called.

allroundexperts commented:
> I can't reproduce locally because I don't know the full steps, but this occurs on builds pointing to prod and we can't merge this PR as is. I'm afraid the best I can do is point to the code which is the most likely cause and share the [stack trace.](https://github.com/Expensify/App/pull/39875#issuecomment-2102646525)

@shubham1206agra Can you reproduce this by any chance?

Julesssss commented:
> @shubham1206agra Can you reproduce this by any chance?

We both weren't able to last time. 

> The alternative was to store the participants in the state

Given it's hard to reproduce, maybe it would be better to try fixing this problem a different way?

allroundexperts commented:
> Given it's hard to reproduce, maybe it would be better to try fixing this problem a different way?

Well, I can't think of any good way. Can either of you please share some ideas?

Julesssss commented:
I don't, but I think it might be a good idea to share the code and exception on our channels to see if anyone has an idea of the bug and possible solution.

allroundexperts commented:
@Julesssss I'll give this another shot today and If still I'm unable to come up with something, I'll post in the channel.

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts committed:
fix: participant list fix

allroundexperts committed:
add spanish translations

allroundexperts commented:
I've pushed a change that should hopefully fix everything. @Julesssss and @shubham1206agra please test now.

Julesssss commented:
I no longer see the crash 👍 

I will update the tests soon to handle all cases in the original issue, but one thing I noticed is that we don't seem to detect the OS permission having been denied on Android: 

- User said continue to our permission modal
- User then denied the native permission modal
- They did this a few times
- Show a 'Denied' modal that explains how users can re-enable the permission manually (similar to this camera permission modal):

![asd](https://github.com/Expensify/App/assets/10736861/2766fd70-811c-42dc-8f53-0dc265ca3885)

Julesssss commented:
I no longer see the crash 👍 

I will update the tests soon to handle all cases in the original issue, but one thing I noticed is that we don't seem to detect the OS permission having been denied on Android: 

- User said continue to our permission modal
- User then denied the native permission modal
- They did this a few times
- Show a 'Denied' modal that explains how users can re-enable the permission manually (similar to this camera permission modal):

<img width="531" alt="Screenshot 2024-05-29 at 13 47 22" src="https://github.com/Expensify/App/assets/10736861/fd194cb5-7f46-4e1e-af70-df8ff0c2a9ee">

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts committed:
type errors

allroundexperts committed:
prettier fix

allroundexperts commented:
@Julesssss When exactly do we need to show that alert? Once the user clicks on the submit button? or when the confirmation screen is opened?

Julesssss commented:
> @Julesssss When exactly do we need to show that alert? Once the user clicks on the submit button? or when the confirmation screen is opened?

Hmm thanks for raising, that is a good question. I think we would only show that in the distance request flow actually but I'd like to handle that separately as a follow up [here](https://github.com/Expensify/App/issues/42978). I simplified test case 2 and will add the distance request test case as a new issue here.

allroundexperts commented:
> * Eventually the OS permission prompt will stop showing, instead our 'Allow camera access' modal should be displayed

Apologies for the confusion. Since you've created a follow up issue for that, is there anything else that's missing in this PR?

allroundexperts committed:
merge with main

Julesssss commented:
Not that I'm aware of 👍 

@shubham1206agra could you please review when you have a chance, thanks!

Julesssss commented:
Thanks. So that sounds to me that we arent caching the permission denied case.

@allroundexperts perhaps in [this](https://github.com/Expensify/App/pull/39875/files#diff-261d3a9129f3b78e4911f17ec1c49bbeed73b010c8c2382f84b3548b51dd1459R53-R59) part of the code or elsewhere we are ignoring the onDeny case? I believe that is the scenario in which we show the pigeon-camera modal -- because at this point users must navigate to settings manually

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts commented:
> I believe that is the scenario in which we show the pigeon-camera modal -- because at this point users must navigate to settings manually

I'm not sure if I am following you here. The pigeon camera modal was for camera permission. Here, we're dealing with location permission only and it does not have any pigeon camera modal.

allroundexperts committed:
Merge branch 'main' into fix-37736

Julesssss commented:
> I'm not sure if I am following you here. The pigeon camera modal was for camera permission. Here, we're dealing with location permission only and it does not have any pigeon camera modal.

Sorry my mistake 🤦 . For this flow I meant the world-pin modal, with a 'Settings' button as the CTA, that is displayed when the OS will no longer allow us to trigger the native prompt -- which is equivalent to the camera-permission flow pigeon.  

Here's the mockup: 

![311315696-837f2141-1461-48e2-a3a3-aea9396600d2](https://github.com/Expensify/App/assets/10736861/24a1d336-8f39-4845-94d8-d50a5482eec5)

allroundexperts committed:
add blocked flow

allroundexperts commented:
@shubham1206agra @Julesssss I've added the denied permission flow. This is ready for review again.

allroundexperts commented:
Also, @Julesssss, can we clarify if we need to show the blocked flow on web? If so, what happens when the user clicks on `Settings` button?

Julesssss commented:
There's a small Podfile issue: `💥 RNPermissions (3.10.1) not found in Podfile.lock. Did you forget to run `npx pod-install`?`

>  can we clarify if we need to show the blocked flow on web? If so, what happens when the user clicks on Settings button?

Another good catch. I think for now it is fine to not trigger any permission flow for web in the blocked flow. By showing the modal we have let them know that they need to enable the permission. 

One thing I noticed is that we don't show any prompt or modal on Desktop. I assume that is because we can't request permissions for the Desktop app?

allroundexperts commented:
> One thing I noticed is that we don't show any prompt or modal on Desktop. I assume that is because we can't request permissions for the Desktop app?

That's correct.

allroundexperts committed:
update pod file

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts committed:
Update Podfile.lock

allroundexperts commented on ios/Podfile.lock:
```diff
@@ -2606,8 +2606,8 @@ SPEC CHECKSUMS:
 PODFILE CHECKSUM: 66a5c97ae1059e4da1993a4ad95abe5d819f555b
 
-COCOAPODS: 1.13.0
+COCOAPODS: 1.14.2
```
Reverted!

allroundexperts commented:
Not sure what's wrong with the podfile. Do you have any clue @shubham1206agra @Julesssss?

allroundexperts committed:
Update Podfile.lock

allroundexperts commented on ios/Podfile.lock:
```diff
@@ -2606,7 +2606,7 @@ SPEC CHECKSUMS:
   Turf: 13d1a92d969ca0311bbc26e8356cca178ce95da2
   VisionCamera: 1394a316c7add37e619c48d7aa40b38b954bf055
-  Yoga: 64cd2a583ead952b0315d5135bf39e053ae9be70
+  Yoga: 1b901a6d6eeba4e8a2e8f308f708691cdb5db312
```
Updated.

Julesssss commented:
A couple of remaining issues: 
- one looks like a new TS conflict  
- one is the RN permissions pod again 😕

Julesssss reviewed (commented):
Nearly there, just resolving a couple of CI issues

allroundexperts committed:
merge with main

allroundexperts committed:
update lock file

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts commented:
All tests passing now @Julesssss @shubham1206agra

Julesssss commented:
Great, taking a look today. @shubham1206agra could you please review again?

allroundexperts commented:
Can you please upload a video of the behaviour?

Julesssss commented:
@allroundexperts here's a video, i think the issues are
- This 'You have to now manually enable permission' modal is being shown out of order. It should be:
    - 1. `We'll use your location to help accurately determine your default currency and timezone. You can edit access in your device's settings anytime.`
    - 2. Native prompt
    - 3. [When native prompt has been denied] `Allowing location access is required to help accurately determine your default currency and timezone. Tap Settings to update permissions.`
- The expense is submitted, instead we should require the user to press the button again when they enable permission

https://github.com/Expensify/App/assets/10736861/0d19783b-942b-48ac-ada1-48732510d576

allroundexperts commented:
> @allroundexperts here's a video, i think the issues are
> 
> * This 'You have to now manually enable permission' modal is being shown out of order. It should be:
>   
>   * 1. `We'll use your location to help accurately determine your default currency and timezone. You can edit access in your device's settings anytime.`
>   * 2. Native prompt
>   * 3. [When native prompt has been denied] `Allowing location access is required to help accurately determine your default currency and timezone. Tap Settings to update permissions.`
> * The expense is submitted, instead we should require the user to press the button again when they enable permission
> 
>  screen-20240617-124325.mp4

@Julesssss I think its showing out of order for you since you denied the permission previously. If you clear your permissions, and then try, it should appear in order. I've also fixed the issue that @shubham1206agra pointed out. Please verify.

allroundexperts committed:
merge with main

Julesssss commented:
> I think its showing out of order for you since you denied the permission previously.

I thought I reset and cleared app data but let me try again this morning 👍

Julesssss commented:
@allroundexperts yeah I retested and both issues are happening on a new install: 


https://github.com/Expensify/App/assets/10736861/ecbb63ec-6715-491c-846c-823e90fa0199

allroundexperts committed:
fix web permissions

allroundexperts committed:
lint issues

allroundexperts commented:
This is a little tricky since each platform has its own set of statuses. I'm still working on it.

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts committed:
WIP: 1777d7ec70 Merge pull request #43140 from Expensify/cmartins-fixButtonWidthDisableTaxSorting

Julesssss commented:
> How should we proceed here? We have the following options:

That sound familiar as we had the [same](https://github.com/Expensify/App/blob/a0a30633cf05d65efd0f07e4ff9fb8cee7463e48/src/pages/iou/request/step/IOURequestStepScan/index.native.tsx#L58-L62) problem for Android camera permissions. 

> 1. For Android, show the alert after the user clicks Continue

That sounds better to me, but could you please just confirm that is the same solution we ended up with?

allroundexperts commented:
@Julesssss For camera permission, we're showing the alert regardless of the platforms. Should we do the same here?

allroundexperts committed:
play sound fix

allroundexperts commented:
@Julesssss, can we also verify the Spanish translations in `src/languages/es.ts` ? I wasn't able to find any existing translations for these and added them as per Google.

Julesssss commented:
> we're showing the alert regardless of the platforms

I know this is slightly different from 'camera' because we auto-prompt instead of having the 'enable camera' button. 

But I'm struggling to imagine how these options are differentfor Android/iOS & others for camera and location. Would you mind explaining the differences and how they look to a user?

Julesssss commented:
> @Julesssss For camera permission, we're showing the alert regardless of the platforms. Should we do the same here?

Yeah, I'll do this after we have fully got the implementation working 👍

allroundexperts commented:
@Julesssss Here are the flows. Is this what you were requesting?

Camera flow:
https://github.com/Expensify/App/assets/30054992/5365603e-08bd-4dfd-98ed-3f4b42567669

Location flow:
https://github.com/Expensify/App/assets/30054992/4a952a6e-bc84-4fe4-adca-e0c1ee90fbb0

allroundexperts commented:
Any update here @Julesssss?

Julesssss commented:
> Here are the flows. Is this what you were requesting?

Yeah, that's helpful. I was trying to remember whether we aligned Android with the other platforms for the camera permission flow. 

I think for now we should do what you suggest here and just show the modals one after another:

> For Android, show the alert after the user clicks Continue button on the first modal and leave the rest of platforms as they are ie

allroundexperts commented:
Sounds good. Thanks. I'll make the changes.

Julesssss commented:
I'm also looking into the translations. I asked for marketing to review the english wording [here](https://github.com/Expensify/App/issues/37736#issuecomment-2194692716). Then I'll get approval for the Spanish translations.

allroundexperts committed:
WIP: 1777d7ec70 Merge pull request #43140 from Expensify/cmartins-fixButtonWidthDisableTaxSorting

allroundexperts committed:
separate android flow from rest of the flows

allroundexperts commented:
@shubham1206agra This is ready for your review again.

Julesssss reviewed (commented):
Copy approved, I have requested Spanish translations. Once complete we can get this PR back in review 👍

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts committed:
prettier fix

Julesssss commented:
> I am seeing `Enable location in Settings` even though I just granted the permission.

Ah, good spot. I guess we need to listen for that event for mweb. If we can close the modal, that is a fine solution IMO. No need to automatically continue with the flow once permission is given

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts committed:
update location timeout

allroundexperts commented:
> Screen.Recording.2024-07-01.at.7.41.39.PM.mov 
> @allroundexperts I am seeing `Enable location in Settings` even though I just granted the permission.

This is fixed!

allroundexperts committed:
settings button update

Julesssss commented:
> @Julesssss @allroundexperts I think there should be a timeout screen (or improve text in settings screen) for web platforms as setting an arbitrary time might not be ideal.

Is the suggestion to show a temporary UI? I don't that is necessary for this flow yet TBH.

Julesssss commented on src/pages/iou/request/step/IOURequestStepScan/LocationPermission/index.ts:
```diff
@@ -0,0 +1,37 @@
+                () => resolve(RESULTS.GRANTED),
+                (error) => resolve(error.TIMEOUT || error.POSITION_UNAVAILABLE ? RESULTS.BLOCKED : RESULTS.DENIED),
+                {
+                    timeout: 5000,
```
COuld you please make this a const, like [this](https://github.com/Expensify/App/blob/6c5d186c3179251d83f1abb437a3e4616a1191c0/src/CONST.ts#L4737).

allroundexperts replied:
Done

allroundexperts committed:
gps timeout constant

Julesssss commented:
I just tested Android again and noticed that we don't prompt for permission in the manual expense flow? We need the same logic to apply to both expense types.

Julesssss commented:
> Why do we need location data for manual expense?

We use provided GPS coordinates on the backend for a things like currency, that are also useful if a manual expense later has a receipt added to it. 

Though to be fair we didn't state this originally, and we don't currently do this on main... I'll follow up with this at a later date 👍 

@shubham1206agra are you happy to check off the review for a final time?

pecanoro commented on src/languages/es.ts:
```diff
@@ -613,6 +614,12 @@ export default {
         cameraErrorMessage: 'Se produjo un error al hacer una foto. Por favor, inténtalo de nuevo.',
+        locationAccessTitle: 'Permitir acceso a la ubicación',
+        locationAccessMessage:
+            'Usaremos su ubicación para determinar con precisión su moneda y zona horaria predeterminadas. Puede editar el acceso en la configuración de su dispositivo en cualquier momento.',
```
```suggestion
            'Usaremos tu ubicación para determinar con precisión la moneda y zona horaria predeterminadas. Puedes editar el acceso en la configuración de tu dispositivo en cualquier momento.',
```

pecanoro commented on src/languages/es.ts:
```diff
@@ -613,6 +614,12 @@ export default {
+        locationAccessTitle: 'Permitir acceso a la ubicación',
+        locationAccessMessage:
+            'Usaremos su ubicación para determinar con precisión su moneda y zona horaria predeterminadas. Puede editar el acceso en la configuración de su dispositivo en cualquier momento.',
+        locationErrorTitle: 'Habilitar ubicación en configuración',
+        locationErrorMessage:
+            'Es necesario permitir el acceso a la ubicación para ayudar a determinar con precisión su moneda y zona horaria predeterminadas. Toque Configuración para actualizar los permisos.',
```
```suggestion
        locationErrorTitle: 'Habilitar ubicación en la configuración',
        locationErrorMessage:
            'Es necesario permitir el acceso a la ubicación para ayudar a determinar con precisión su moneda y zona horaria predeterminadas. Haz click en Configuración para actualizar los permisos.',
```

pecanoro reviewed (dismissed):
Not sure if you noticed, but there are some really close translations here too: https://github.com/Expensify/App/blob/f220460735429186fa8c4996351fff83936fc5f7/src/languages/es.ts#L356-L362

allroundexperts committed:
Update src/languages/es.ts

Co-authored-by: Rocio Perez-Cano <pecanoro@users.noreply.github.com>

allroundexperts committed:
Update src/languages/es.ts

Co-authored-by: Rocio Perez-Cano <pecanoro@users.noreply.github.com>

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts commented:
@shubham1206agra Merged main.

pecanoro commented on src/components/LocationPermissionModal.tsx:
```diff
@@ -0,0 +1,104 @@
+        cb();
+    };
+
+    const onConfirm = errorHandler(() => {
```
I am going to leave a comment here but this applies everywhere. These don't seem to be following the convention,  these should be named for what it does, not what it handles: https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#event-handlers

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts committed:
rename functions to follow conventions

allroundexperts commented:
> @allroundexperts Can you fix lint please?

All done!

allroundexperts committed:
Merge branch 'main' into fix-37736

allroundexperts committed:
Merge branch 'main' into fix-37736

Julesssss commented:
Excellent, the flow is working well. I will create new test rail cases for these flows after merging.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9946176278) to staging by https://github.com/Julesssss in version: 9.0.7-3 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Cherry-picked](https://github.com/Expensify/App/actions/runs/9947036855) to staging by https://github.com/thienlnam in version: 9.0.7-4 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

@Expensify/applauseleads please QA this PR and check it off on the [deploy checklist](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3AStagingDeployCash) if it passes.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9965822856) to production by https://github.com/thienlnam in version: 9.0.7-8 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

Julesssss commented:
Hey @allroundexperts. This isn't a big problem, but [this small issue](https://github.com/Expensify/App/issues/45431) was noticed in the flow (settings button not opening location permission settings). Would you mind taking a look?

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9992557463) to production by https://github.com/mountiny in version: 9.0.8-6 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
